plugins/inputs/opensearch_query/opensearch_query.go

```diff
@@ -0,0 +1,326 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package opensearch_query
+
+import (
+	"context"
+	"crypto/tls"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	influxtls "github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/opensearch-project/opensearch-go/v2"
+	"github.com/opensearch-project/opensearch-go/v2/opensearchapi"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+// OpensearchQuery struct
+type OpensearchQuery struct {
+	URLs                []string        `toml:"urls"`
+	Username            string          `toml:"username"`
+	Password            string          `toml:"password"`
+	EnableSniffer       bool            `toml:"enable_sniffer"`
+	Timeout             config.Duration `toml:"timeout"`
+	HealthCheckInterval config.Duration `toml:"health_check_interval"`
+	Aggregations        []osAggregation `toml:"aggregation"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	influxtls.ClientConfig
+	osClient *opensearch.Client
+}
+
+// osAggregation struct
+type osAggregation struct {
+	Index             string          `toml:"index"`
+	MeasurementName   string          `toml:"measurement_name"`
+	DateField         string          `toml:"date_field"`
+	DateFieldFormat   string          `toml:"date_field_custom_format"`
+	QueryPeriod       config.Duration `toml:"query_period"`
+	FilterQuery       string          `toml:"filter_query"`
+	MetricFields      []string        `toml:"metric_fields"`
+	MetricFunction    string          `toml:"metric_function"`
+	Tags              []string        `toml:"tags"`
+	IncludeMissingTag bool            `toml:"include_missing_tag"`
+	MissingTagValue   string          `toml:"missing_tag_value"`
+	mapMetricFields   map[string]string
+
+	aggregation AggregationRequest
+}
+
+type mapping map[string]fieldIndex
+
+type fieldIndex struct {
+	Mappings map[string]fieldMapping `json:"mappings"`
+}
+
+type fieldMapping struct {
+	FullName string               `json:"full_name"`
+	Mapping  map[string]fieldType `json:"mapping"`
+}
+
+type fieldType struct {
+	Type string `json:"type"`
+}
+
+func (*OpensearchQuery) SampleConfig() string {
+	return sampleConfig
+}
+
+// Init the plugin.
+func (o *OpensearchQuery) Init() error {
+	if o.URLs == nil {
+		return fmt.Errorf("no urls defined")
+	}
+
+	err := o.connectToOpensearch()
+	if err != nil {
+		o.Log.Warnf("error connecting to opensearch: %w", err)
+	}
+
```

Comment:
You should **NOT** connect to an external service in `Init()`! Please use the `Start`/`Stop` service interface instead.
Furthermore, if the connect fails here, I see no recover mechanism anywhere, i.e. you should fail here and not just warn for an non-recoverable error.

Reply:
Connecting can be done  to be able to warn users, but indeed, a retry mechanism seems missing in the Gather cycle?

Reply:
I've refactored the code to remove the initial call to the cluster to fetch the mappings.  I think I was originally matching what the `elasticsearch_query` was doing, but all the error cases are handled properly without needing to make this check.

As for the retry mechanism: the OpenSearch client handles retries internally (3 by default).  If we're unable to connect to the cluster even after these retries, we'll see the error in the logs with the explanation.

Reply:
Sven means if we cannot connect when Telegraf is started, but for example the cluster comes up a few minutes after that, Telegraf will currently never be able to retrieve data from it..

Reply:
As I mentioned earlier, I've refactored the initial connection part of out `Init()`, since it wasn't actually necessary.  The first time Telegraf connects to the OpenSearch cluster is on `Gather()`.

Reply:
@Hipska I think we should move all "external connections" to be made in start. We can then work on implementing a general retry mechanism in Telegraf... Something that is on my todo list (which is long unfortunately).

Reply:
Isn't start only for service plugins?